modules/system-disaster-recovery/src/main/java/org/apache/ignite/internal/disaster/system/MetastorageRepairImpl.java

```diff
@@ -0,0 +1,181 @@
+
+                if (contains(cumulativeValidatedNodeNames, nodeNames)) {
+                    future.complete(null);
+                }
+            }
+
+            @Override
+            public void onNodeInvalidated(LogicalNode invalidatedNode) {
+                cumulativeValidatedNodeNames.remove(invalidatedNode.name());
+            }
+        };
+
+        logicalTopology.addEventListener(listener);
+
+        cmgManager.validatedNodes()
+                .thenAccept(validatedNodes -> {
+                    Set<String> validatedNodeNames = validatedNodes.stream()
+                            .map(ClusterNode::name)
+                            .collect(toSet());
+                    if (contains(validatedNodeNames, nodeNames)) {
+                        future.complete(null);
+                    }
+
+                    cumulativeValidatedNodeNames.addAll(validatedNodeNames);
+                });
+
+        return future
+                .thenRun(() -> logicalTopology.removeEventListener(listener));
+    }
+
+    private static boolean contains(Set<String> container, Set<String> containee) {
+        return difference(containee, container).isEmpty();
+    }
+
+    private CompletableFuture<Map<String, IndexWithTerm>> collectMetastorageIndexes(Set<String> participatingNodeNames) {
+        MetastorageIndexTermRequestMessage request = messagesFactory.metastorageIndexTermRequestMessage().build();
+
+        Map<String, CompletableFuture<MetastorageIndexTermResponseMessage>> responses = new HashMap<>();
+
+        for (String nodeName : participatingNodeNames) {
+            responses.put(
+                    nodeName,
+                    messagingService.invoke(nodeName, request, 10_000).thenApply(MetastorageIndexTermResponseMessage.class::cast)
+            );
+        }
+
+        return allOf(responses.values()).thenApply(unused -> {
+            return responses.entrySet().stream()
+                    .collect(toMap(Entry::getKey, entry -> indexWithTerm(responses.get(entry.getKey()).join())));
+        });
+    }
+
+    private static IndexWithTerm indexWithTerm(MetastorageIndexTermResponseMessage message) {
+        return new IndexWithTerm(message.raftIndex(), message.raftTerm());
+    }
+
+    private static Set<String> nodesWithBestIndexes(Map<String, IndexWithTerm> indexes, int metastorageReplicationFactor) {
+        return indexes.entrySet().stream()
+                .sorted(Entry.<String, IndexWithTerm>comparingByValue().reversed())
+                .limit(metastorageReplicationFactor)
+                .map(Entry::getKey)
+                .collect(toSet());
+    }
+
+    private static String chooseNodeWithBestIndex(Map<String, IndexWithTerm> indexes, Set<String> newMgNodes) {
+        return newMgNodes.stream()
+                .sorted(Comparator.<String, IndexWithTerm>comparing(indexes::get).reversed())
+                .limit(1)
+                .findFirst().orElseThrow();
```

Comment:
There's a `Stream#min` and `Stream#max`, depending on the desired comparison order. I suggest using one of these methods